pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/state/BKStateStoreImpl.java

```diff
@@ -190,4 +191,39 @@ public ByteBuffer get(String key) {
             throw new RuntimeException("Failed to retrieve the state value for key '" + key + "'", e);
         }
     }
+
+    @Override
+    public StateValue getStateValue(String key) {
+        try {
+            return result(getStateValueAsync(key));
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to retrieve the state value for key '" + key + "'", e);
+        }
+    }
+
+    @Override
+    public CompletableFuture<StateValue> getStateValueAsync(String key) {
+        return table.getKv(Unpooled.wrappedBuffer(key.getBytes(UTF_8))).thenApply(
+                data -> {
+                    try {
+                        if (data != null && data.value() != null) {
+                            ByteBuffer result = ByteBuffer.allocate(data.value().readableBytes());
+                            data.value().readBytes(result);
+                            // Set position to off the buffer to the beginning, since the position after the
+                            // read is going to be end of the buffer
+                            // If we do not rewind to the beginning here, users will have to explicitly do
+                            // this in their function code
+                            // in order to use any of the ByteBuffer operations
+                            result.position(0);
```

Comment:
isn't this making the assumption that position was 0 before reading? Aren't ByteBuffer's `mark` and `reset` meant to be used for this type of use cases?

Reply:
got it now, I just used the same code in line: https://github.com/apache/pulsar/blob/b56efff3f125b107d39eb7a88c040badc721f692/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/state/BKStateStoreImpl.java#L179

Reply:
but the result is a new init `ByteBuffer`, I think its position was 0 before reading?